### ISSUE TITLE
Add an electronic signature to a PDF

### ISSUE THREAD

**Issue by PrSunflower** [sumatrapdfreader/sumatrapdf#1744]
Hi, thank you very much to all devs, Sumatra is a great PDF reader and has replaced Adobe Acrobat for me... except for one feature! 😄

It would be great if it was possible to add an **electronic signature** (not a [cryptographic digital signature](https://www.howtogeek.com/164668/how-to-electronically-sign-documents-without-printing-and-scanning-them/)) just like the [**"Fill & Sign" tool**](https://www.jotform.com/blog/how-to-add-a-signature-to-a-pdf/) in Acrobat Reader.

Basically I want to add a PNG picture of my signature on top of an existing PDF document.

**Comment by GitHubRulesOK** [comment]
@kjk 
in some apps this is most easily done using the annotation features with a custom stamp. there **was** <strike>an open issue https://github.com/sumatrapdfreader/sumatrapdf/issues/1681 that when moving custom stamps they get destroyed, that would need to be fixed first.</strike> an issue now closed but not the only problem with using custom stamps.

**LATER EDIT**
This is possible in 3.7 pre-release where a RUN script can place an image a box and text all in one command. clearly to do all that in one line it uses switches with a JS script the script is at https://github.com/GitHubRulesOK/SumatraPDF-Plus/tree/master/Scripts#addoverlayjs

<img width="362" height="368" alt="Image" src="https://github.com/user-attachments/assets/4c2a003e-9038-48c1-96a7-1ad3aa3a6ec4" />

```
"C:\Program Files\SumatraPDF\SumatraPDF.exe"  run "AddOverlay.js" -p=1 -r=15 -bs=0 -b=300,250,200,110 -img="Picasso.png" -to=86,-35  -t="  Welcome,\n\n   Signature\n To my World!" "AddOverlay\input.pdf" -to=20,120 -bs=0.4

```

**Comment by metya** [comment]
Are there any plans for that feature?)

**Comment by GitHubRulesOK** [comment]
which feature ? 
Adding an image to a PDF is trivial for many powerful Editors, but is not a native feature in MuPDF-GL reader It will need someone to write  an editor function just to stamp a static png of my signature.

For significantly complex secured electronic signatures needs a team as big as the one amongst the tens of thousands employed by Adobe

**Comment by PrSunflower** [comment]
@GitHubRulesOK thanks for the precision 👌

> For significantly complex secured electronic signatures as raised by OP it needs a team as big as the one amongst the tens of thousands employed by Adobe

Yeah that's why I asked for **electronic signatures**, not **digital signatures**. I cannot ask that to a team of voluntary developers. <br>
Moreover implementing digital signatures would require validation from certificate authorities and I am not sure if that's compatible with the developers' vision for their free & open-source software.

_Side note: What a coincidence it's been exactly 1 year since I've created this issue!_

**Comment by GitHubRulesOK** [comment]
Personally I have a low regard of the value of any means of signing, unless there is a form of [Escroue](https://en.wikipedia.org/wiki/Escrow) which is basically what is provided by Full Witnessed Certified Digital Signatures (and even then are open to abuse)

![image](https://user-images.githubusercontent.com/12754913/131880813-057ec296-fc3b-4aff-a2e2-a464462e7f71.png)

**Comment by fkrivsky** [comment]
This would be super useful feature. 
I've been signing a lot of documents lately (since Covid basically) using Foxit Reader and this feature missing in Sumatra is a deal breaker for me...

**Comment by GitHubRulesOK** [comment]
The common signature stamping of pages (just like running through a bulk rubber stamping machine) can be done at 1000's pages per minute by any command line tool (there is no need for a GUI, if the page position is the same. Likewise it is a waste of time rubber stamping one or more pages to approve actions since an e-mail "my word is my bond" saying "sanctioned" (let it happen) should be good enough. the key is 3rd party holding the deed. *1590s, in law, "a writing fully executed by the parties, but put into the custody of a third person to hold until the fulfilment of some condition,*

**Comment by fkrivsky** [comment]
> The common signature stamping of pages (just like running through a bulk rubber stamping machine) can be done at 1000's pages per minute by any command line tool (there is no need for a GUI, if the page position is the same.) Likewise it is a waste of time rubber stamping one or more pages to approve actions since an e-mail "my word is my bond" saying "sanctioned" (let it happen) should be good enough. the key is 3rd party holding the deed. _1590s, in law, "a writing fully executed by the parties, but put into the custody of a third person to hold until the fulfilment of some condition,_ Like death will and in pdf terms that requires an electronic locked (encrypted) witness service.

In my country we don't use 1590s laws.
Also I meant like 50 different documents in the last couple of years, not the best choice of words I see :) A "lot" comparatively to the past... 
I just need to review a document and place the signature stamp manually wherever needed, which is a common practice here.

**Comment by llewynS** [comment]
> Personally I have a low regard of the value of any means of signing, unless there is a form of [Escroue](https://en.wikipedia.org/wiki/Escrow) which is basically what is provided by Full Witnessed Certified Digital Signatures (and even then are open to abuse)
> 
> ![image](https://user-images.githubusercontent.com/12754913/131880813-057ec296-fc3b-4aff-a2e2-a464462e7f71.png)

Your low regard is irrelevant to the request. A custom stamp that includes ones signature is a basic feature that is useful in most countries. Not having the feature would reduce the number of people that find this program useful, e.g I just downloaded it purely for this and have no use for it without this feature. Being able to place your signature on a pdf removes the need to print, sign, and rescan. A personal signature is not secure, so it doesn't reduce validity, it does reduce waste, and it is more convenient. Your country is extremely officious if they don't allow this.

**Comment by GitHubRulesOK** [comment]
@llewynS without laboring the point I got paid well for injecting my image and the image of checker and line manager and his boss on every document I produced, because i was paid less than them. the real method of verification was i sent documents through a chain that they should have actioned the bulk receipt of and not rejected.

**Comment by caramdache** [comment]
I am also in need of being able to insert an image into a PDF document. The document types vary, so it's not able to set programmatically (I wish it was).

Not having this feature is also holding me back from switching totally from Adobe Reader or Foxit to Sumatra PDF.

**Comment by GitHubRulesOK** [comment]
I use Tracker exchange as lighter than Foxit and it's easy to set both to jump from one to the other on same page without closing and opening either.
![image](https://user-images.githubusercontent.com/12754913/191335709-7b260587-1f8b-431d-b278-b52327101466.png)

**Comment by caramdache** [comment]
> I use Tracker exchange as lighter than Foxit and it's easy to set both to jump from one to the other on same page without closing and opening either. 

Excellent indeed!

Here's how to do it, just in case anyone comes around here:

1. Install `Tracker Exchange` (now `PDF-XChange Editor').
2. Open a PDF document in SumaryPDF.
3. File/Open in PDF-XChange
4. Edit the document in PDF-XChange, then File/Save
5. SumatraPDF refreshes automatically and displays the update PDF.

**Comment by GitHubRulesOK** [comment]
@caramdache and others
You can do the same with Edge inking or any free text overwrite, beware Edge uses an unconventional HTML/Word means to scale text which MuPDF/SumatraPDF will revert to conventional PDF spaced plainer text. The bonus is easier to add up to 256 chars in many languages, however avoid using more than 256 in any comments (SumatraPDF has had to apply a limitation).

the point about open one in the other is in exchange using a customised "launch" sumatra pdf with filename and page ID and conversely in SumatraPDF an ExternalViewers command passing filename and page number.

**Comment by polar147** [comment]
The MuPDF disponibilize the mutool that can digital sign PDF files

https://www.mupdf.com/docs/manual-mutool-sign.html

**Comment by AlexErrant** [comment]
Firefox can do some basic edits to pdf https://www.mozilla.org/en-US/firefox/features/pdf-editor/

**Comment by GitHubRulesOK** [comment]
there is nothing really special about signing just the ease with which it may be added and removed (PDF is no more secure signed than unsigned) and thus signatures are simply added and removed just like text when coded/decoded.

here a text file with a PDF extension and a signing box in SumatraPDF can invoke Adobe signing abd read the returned signed file. that digital signature is acceptable in law and deletable too!

![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/72038a1c-5d72-4851-933d-d8538ffcc046)

**Comment by thomasdn** [comment]
If it would be possible to add a simple signature image overlay on a PDF that would be highly appreciated. Acrobat Reader supports this basic way of signing PDFs. While this is not a secure cryptograpic signature or similar, it is good enough for lots of use cases.

**Comment by GitHubRulesOK** [comment]
@thomasdn adding a signature electronically is already possible with Windows Edge you can send a PDF direct from SumatraPDF to be directly opened there signed and see the signature in SumatraPDF window on saving file.

![Image](https://github.com/user-attachments/assets/e21f466f-30c1-4809-9525-7f40f78beefc)

![Image](https://github.com/user-attachments/assets/b4f733c5-7219-4e4a-b70a-28c624b5f6f4)

Currently SumatraPDF cannot alter that much (no longer allows drag which was an earlier problem) but can change colour or line thickness.

![Image](https://github.com/user-attachments/assets/2356cf16-344d-40f0-8e27-3d0a4af70f81)

**Comment by GitHubRulesOK** [comment]
To state the obvious there are many ways to add images or ink annotations to PDFS  that are provided by the Native Windows Apps 

However to answer the OP initial question I wrote a multipurpose script that can thus for any user be customised to space a transparent image onto a pdf page so In this space you can see my artwork. It has too many abilities to describe them all but basic need is
```
"C:\Program Files\SumatraPDF\SumatraPDF.exe"  run "AddOverlay.js" -p=1 -img="Picasso.png" "..\..\..\..\..\..\Downloads\Add an electronic signature.pdf" -b=375,530,100,55

-p=page number -img="filename of a transparent.PNG or .JPG" there are several ways to adjust position but in this case
-b=375,530,100,55 means in points units X=375 , Y from PDF datum =530 the image desired width (related to pixels ratio) =100
and a height of 55 points (this is a ratio of double normal density e.g. 2 pixels per point) 

RAW images are 96dpi nominal but there are no DPI in a PDF thus we set the number of dots per points length 
Thus if 200 dots are to be placed across 100 points that is **similar to** but NOT 144 DPI as there are other scaler effects.
```
<img width="1086" height="558" alt="Image" src="https://github.com/user-attachments/assets/3cf9f8ca-bca7-4adb-83f2-d5225374e5d5" />

see also the example above https://github.com/sumatrapdfreader/sumatrapdf/issues/1744#issuecomment-691552655

script for placing images via command is at https://github.com/GitHubRulesOK/SumatraPDF-Plus/tree/master/Scripts#addoverlayjs

There is now a floating tool more suited to placing transparent signature by click on the location to be placed so see  
https://github.com/GitHubRulesOK/SumatraPDF-Plus#putimage-new

**Comment by GitHubRulesOK** [comment]
OK for those that insist on digital signatures I have provided a plugin script to place images with signature or add to a digital e-signing the script is a first attempt so may be raw at the edges but works and can be easily edited to your preferences.
Download from top right of https://github.com/GitHubRulesOK/SumatraPDF-Plus/blob/master/PlaceSig.cmd

<img width="1219" height="400" alt="Image" src="https://github.com/user-attachments/assets/3b059125-820d-445c-92bc-9eb098872040" />

<img width="1039" height="543" alt="Image" src="https://github.com/user-attachments/assets/beb4ea26-7f0d-4427-b53e-48b34b6bb247" />
